doc/getnote.1
.TH GETNOTE 1 "August 2004" "Newts" "Newts Reference Manual"

.SH NAME
getnote \- print a single note or response to standard output

.SH SYNOPSIS
.B getnote
[\fIoptions\fR] \fINOTESFILE\fR \fINOTE\fR [\fIRESPONSE\fR]

.SH DESCRIPTION
.B getnote
prints either a single note or a single response from a particular notesfile to
standard output.

.SH OPTIONS

.TP
\fB\-h\fR, \fB\-\^\-help\fR
Print a summary of usage and command-line options for
.B getnote
and exit.

.TP
\fB\-p\fR, \fB\-\^\-print\-header\fR
Preface the text of the note with a header containing information about the
note, including notesfile, note and response number, title, director message,
and author.

.TP
\fB\-\^\-debug\fR
Print debugging messages to standard error.

.TP
\fB\-\^\-version\fR
Print version information for
.B getnote
and exit.

.SH AUTHOR
.B Newts
was written and is maintained by Tyler Berry <tyler+newts@thoughtlocker.net>.

.SH SEE ALSO
\fBautoseq\fR(1), \fBchecknotes\fR(1), \fBmknf\fR(1), \fBnfadmin\fR(1),
\fBnfdump\fR(1), \fBnfload\fR(1), \fBnfmail\fR(1), \fBnfpipe\fR(1),
\fBnfprint\fR(1), \fBnfstats\fR(1), \fBnftimestamp\fR(1), \fBnotes\fR(1),
\fBrmnf\fR(1)

The full documentation for
.B Newts
is maintained as a Texinfo manual.  If the
.B info
and
.B Newts
programs are properly installed at your site, the command
.IP
.B info newts
.PP
should give you access to the complete manual.


NEWS
Newts NEWS                                                         -*- Text -*-
-------------------------------------------------------------------------------

* ______ - Newts 0.15

- More unit tests.
- More changes and additions to the client API.
- More documentation of the client API.

* Wednesday, December 14, 2005 - Newts 0.14.8

- Include the SVN revision number in the output of --version.
- Added unit tests for the client API.
- Began adding documentation for the client API in a Doxygen-compatible way.
- Substantial changes and additions to the client API.
- Substantial refactoring in the main client, both to comply with the revised
  client API and to clean the code up in general.
- Modified the utility frontend programs to use the revised client API.
- Refactored the pager to be modular, for purposes of future rewrites to add
  things like Unicode support.
- Use gnulib-tool to generate Gnulib support.
- Automatically use the most recent version of Libtool.

* Tuesday, July 26, 2005 - Newts 0.14.7

- Fixed signal handling so that suspending a spawned process works correctly.
- Added David Mellis's python bindings to the repository and distribution.

* Friday, June 16, 2005 - Newts 0.14.6

- Fixed a bug in nfdump where the last note in a dumped notesfile was omitted.
- Added 'g' to perform one-touch seqing.

* Monday, March 21, 2005 - Newts 0.14.5

- Updated to Libtool 1.5.14.
- Fixed a segfault when trying to use the alternate sequencer.
- Made the alternate sequencer activate the sequencer automatically.
- Made the alternate sequencer work correctly with the '--user' option.
- Fixed the broken --time option, which was not actually setting the time at
  all. Oops.
- Fixed a few odd display bugs.
- Display messages are now 'Unicode-friendly'.
  See <http://www.cl.cam.ac.uk/~mgk25/ucs/quotes.html> for details.
- Resynced with Gnulib.

* Sunday, January 23, 2005 - Newts 0.14.4

- Updated to Libtool 1.5.10.
- Corrected some mistakes in the Texinfo manual.
- Fixed a bug which would cause a segfault when replying to a note from a local
  user who had been deleted from the system, or forwarding a mail from a note
  written by such a user.
- Fixed a bug in nfload where the dumpfile would fail to open if nfload was run
  as root. (It was dropping root permissions and not picking them back up.)
- Resynced with Gnulib, which involved some fairly major updates to xmalloc and
  friends, as the interface Gnulib uses changed.

* Tuesday, September 14, 2004 - Newts 0.14.3

- Corrected a typo in nfadmin.1.
- Synchronized with new files in Gnulib.

* Tuesday, August 10, 2004 - Newts 0.14.2

- Corrected a Whole Bunch of inconsistencies between the Texinfo manual, the
  newly-developed manpages, and the actual operation of the programs.
- Added manpages for all currently-built programs.
- Renamed 'printnote' to 'getnote'.
- Removed the '-d' short option from all programs, which used to correspond to
  '--debug'.  It was blocking the use of '-d' for more important options which
  are far more likely to see use.

* Tuesday, June 29, 2004 - Newts 0.14.1

- Completely redid all compatibility code using a recent CVS version of
  Gnulib.

* Thursday, June 10, 2004 - Newts 0.14

- Adding the missing code to the nfload program which will load access entries
  from the dumpfile into the notesfile.
- Added debugging support to all of the various binaries built by the package,
  via the --debug command-line option.
- Added an option for debugging that allows the notes client to setuid to an
  arbitrary user when being run by root. Useful for running notes inside gdb.
- Changed the license for the Newts Reference Manual - it's no longer GFDL.
- Fixed a bug which would cause a segfault when reading a note from a local
  user who had been deleted from the system.
- Reorganized the entire directory structure of the source tree.
- Reorganized header files to install the relevant interfaces.
- Began systematic documentation of the libnewtsclient interface.

* Tuesday, March 9, 2004 - Newts 0.13.2

- Fixed a display bug which caused informational messages and prompts to fail
  to display when a person tried to enter a notesfile without rights.
- Syncronized with new files in Gnulib.

* Friday, January 16, 2003 - Newts 0.13.1

- Synchronized with new releases of GNU Autotools and Gnulib.

* Tuesday, December 9, 2003 - Newts 0.13

- Added nfdump, completing the set of basic utilities.
- Added some sanity checks to the 't' talk command.
- Various small bits of code cleanup.
- Made sure that root doesn't accidentally lose the ability to write files.

* Monday, September 29, 2003 - Newts 0.12.7

- Portability work for sparc-sun-solaris2.6.
- utmp scanning to determine whether an author is online.
- Synchronization and updates for portability code.

* Thursday, September 18, 2003 - Newts 0.12.6

- Added a couple missing m4 files.

* Friday, September 5, 2003 - Newts 0.12.5

- Code cleanup, fixes, and error-checking.

* Saturday, August 30, 2003 - Newts 0.12.4

- Added man pages for notes(1) and autoseq(1).
- Fixed inconsistencies between documentation and actual behavior.
- Made a couple command line options make more sense.
- Fixed distribution of a couple missing files.

* Tuesday, August 26, 2003 - Newts 0.12.3

- Fixed a mistake with an earlier fix to the 'N' key.
- Prepared for a CVS branch to add Readline support.

* Tuesday, August 19, 2003 - Newts 0.12.2

- Interim release for code maintenance.
- Lots and lots of code cleanup.
- Fixed lots of FIXMEs in the code; these were largely for incorrect usage of
  one kind or another; static buffers instead of dynamic alloc, not checking
  return values, etc. It's not there yet, but it's getting there.

* Wednesday, August 13, 2003 - Newts 0.12.1

- Staving off disaster, a bug in notesfile parsing was discovered and slain in
  the embryonic version 0.12, which is supplanted by this release.
- Finally fixed that nfload bug that's been driving me nuts for the past three

clients/frontends/getnote.c
/*
 * getnote.c - print single notes to stdout
 *
 * This file is part of the Newts notesfile system.
 * Copyright (C) 2003, 2004, 2005 Tyler Berry.
 *
 * Newts is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later
 * version.
 *
 * Newts is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Newts; if not, write to the Free Software Foundation, Inc., 59 Temple Place,
 * Suite 330, Boston, MA 02111-1307 USA
 */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#include "frontend.h"

#include "dirname.h"
#include "error.h"
#include "getopt.h"
#include "strpbrk.h"

#if HAVE_FCNTL_H
# include <fcntl.h>
#endif

#if HAVE_PWD_H
# include <pwd.h>
#endif

#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

/* Whether to display debugging messages. */
int debug = FALSE;

/* Whether to print a header. */
int print_header = FALSE;

int
main (int argc, char **argv)
{
  newts_nfref *ref;
  struct notesfile nf;
  struct newt note;

  int opt;
  int option_index = 0;
  extern char *optarg;
  extern int optind, opterr, optopt;

  struct option long_options[] =
    {
      {N_("debug"),0,0,'D'},
      {N_("help"),0,0,'h'},
      {N_("print-header"),0,0,'p'},
      {N_("version"),0,0,0},
      {0,0,0,0}
    };

  memset (&nf, 0, sizeof (struct notesfile));
  memset (&note, 0, sizeof (struct newt));

#ifdef __GLIBC__
  program_name = program_invocation_short_name;
#else
  program_name = base_name (argv[0]);
#endif

  /* Initialize i18n. */

#ifdef HAVE_SETLOCALE
  setlocale (LC_ALL, "");
#endif

#if ENABLE_NLS
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);
#endif

  setup ();

  while ((opt = getopt_long (argc, argv, N_("hp"),
                             long_options, &option_index)) != -1)
    {
      switch (opt)
        {
        case 0:
          {
            printf_version_string (N_("getnote"));

            teardown ();

            if (fclose (stdout) == EOF)
              error (EXIT_FAILURE, errno, _("error writing output"));
            exit (EXIT_SUCCESS);
          }

        case 'D':
          debug = TRUE;
          break;

        case 'p':
          print_header = TRUE;
          break;

        case 'h':
          printf (_("Usage: %s [OPTION]... NOTESFILE NOTENUM [RESPNUM]\n"
                    "Print the specified note from NOTESFILE.\n\n"),
                  program_name);

          printf (_("  -p, --print-header   Print header information (author, etc.) for the note\n"
                    "  -h, --help           Display this help and exit\n"
                    "      --debug          Display debugging messages\n"
                    "      --version        Display version information and exit\n\n"));

          printf (_("Report bugs to <%s>.\n"), PACKAGE_BUGREPORT);

          teardown ();

          if (fclose (stdout) == EOF)
            error (EXIT_FAILURE, errno, _("error writing output"));
          exit (EXIT_SUCCESS);

        case '?':
          fprintf (stderr, _("Try '%s --help' for more information.\n"),
                   program_name);

          teardown ();

          exit (EXIT_FAILURE);
        }
    }

  if (optind == argc || optind + 1 == argc)
    {
      fprintf (stderr, _("%s: too few arguments\n"), program_name);
      fprintf (stderr, _("Try '%s --help' for more information.\n"),
               program_name);

      teardown ();

      exit (EXIT_FAILURE);
    }
  else if (optind + 2 == argc) /* NF NOTENUM */
    {
      ref = nfref_alloc ();
      parse_single_nf (argv[optind++], ref);
      note.nr.notenum = atoi (argv[optind]);
      note.nr.respnum = 0;
    }
  else if (optind + 3 == argc) /* NF NOTENUM RESPNUM */
    {
      ref = nfref_alloc ();
      parse_single_nf (argv[optind++], ref);
      note.nr.notenum = atoi (argv[optind++]);
      note.nr.respnum = atoi (argv[optind]);
    }
  else
    {
      fprintf (stderr, _("%s: too many arguments\n"), program_name);
      fprintf (stderr, _("Try '%s --help' for more information.\n"),
               program_name);

      teardown ();

      exit (EXIT_FAILURE);
    }

  if (open_nf (ref, &nf) != NEWTS_NO_ERROR)
    {
      nfref_free (ref);
      teardown ();

      error (EXIT_FAILURE, 0, _("error opening notesfile '%s'"),
             nfref_pretty_name (nf.ref));
    }

  if (!(nf.perms & READ) && !(nf.perms & DIRECTOR))
    {
      nfref_free (ref);
      teardown ();

      error (EXIT_FAILURE, 0, _("you are not allowed to read notesfile '%s'"),
             nfref_pretty_name (nf.ref));
    }

  nfref_copy (&note.nr.nfr, nf.ref);
  if (get_note (&note, TRUE) >= 0)
    {
      if (print_header)
        {
          struct passwd *pw = getpwnam (note.auth.name);

          printf (_("Notesfile: %s\n"), nfref_pretty_name (nf.ref));

          printf (_("Note: %d of %d\n"), note.nr.notenum, nf.total_notes);
          if (note.nr.respnum != 0)
            printf (_("Response: %d of %d\n"), note.nr.respnum, note.total_resps);
          printf (_("Title: %s\n"), note.title);
          if (note.director_message)
            printf (_("Director message: %s\n"), note.director_message);

          if (pw)
            {
              char *real = newts_strdup (pw->pw_gecos);
              char *separator;

              if ((separator = strpbrk (real, N_(":,"))) != NULL)
                *separator = '\0';

              printf (_("Author: %s <%s@%s>\n"), real, note.auth.name,
                      note.auth.system);

              newts_free (real);
            }
          else
            printf (_("Author: %s@%s\n"), note.auth.name, note.auth.system);

          putchar ('\n');
        }

      printf (N_("%s"), note.text);
    }
  else
    {
      fprintf (stderr, _("%s: no such note\n"), program_name);

      nfref_free (ref);
      teardown ();

      exit (EXIT_FAILURE);
    }

  nfref_free (ref);
  teardown ();

  if (fclose (stdout) == EOF)
    error (EXIT_FAILURE, errno, _("error writing output"));

  exit (EXIT_SUCCESS);
}
